### ISSUE TITLE
Relative Links when Exporting to HTML No Longer Working

### ISSUE THREAD

**Issue by adamkingsbury** [jgraph/drawio#2840]
Hi I have recently discovered that in some of the latest builds of Draw.io, when exporting a document to html, referencing external documents as links via relative references no longer works. This had previously worked (we had built process maps with cross references to HTML work instructions and packaged the exported HTML and sub-folders containing the work instructions on a network drive successfully)

Process:

1) Build a destination structure like so:

Root Folder
|
|-- processmap.drawio
|-- processmap.html           (the exported html file)
|-- workinstructions            (a sub-folder)
   |-- workinstruction.html  (an html file in the sub-folder)

2) in processmap.drawio, add a link to a node. As the destination use a relative reverence for "/workinstructions/workinstruction.html"

3) export processmap to HTML
4) open html in browser
5) click on node in browser.

Previously the above would have opened the link either in same window or new tab. Now nothing happens

**Comment by adamkingsbury** [comment]
So far I've discovered that only full Urls will work.

"www.google.com" will not work
"http://www.google.com" does work

Only workaround I can think of is actually hosting the files you want to link to on a Web server so that you can actually reference a full Web address, but that's not workable in some of my situations.

**Comment by cgulati001** [comment]
Yes I have the same issue in that complete URLs are working fine but my documents require the embedded links.

**Comment by adamkingsbury** [comment]
Doing some further digging.....it looks like it is an issue with the allowable URIs in the sanitize function in DOMPurify

when clicking a relative link such as "folder/anotherfile.html" if you've opened the exported html from local disk, the resolved URI will actually be something like **"file:///C:/folder/anotherfile.html"**

From the documentation, the following regex is what DOMPurify allows by default:
```
// allow specific protocols handlers in URL attributes via regex (default is false, be careful, XSS risk)
// by default only http, https, ftp, ftps, tel, mailto, callto, cid and xmpp are allowed.
// Default RegExp: /^(?:(?:(?:f|ht)tps?|mailto|tel|callto|cid|xmpp):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i;
```

the example URI above will fail as the file protocol is not permitted

if the regex was customised to also allow the "file" protocol, I think the issue would be resolved.


the example from the DOMPurify website:

```
// allow specific protocols handlers in URL attributes via regex (default is false, be careful, XSS risk)
// by default only http, https, ftp, ftps, tel, mailto, callto, cid and xmpp are allowed.
// Default RegExp: /^(?:(?:(?:f|ht)tps?|mailto|tel|callto|cid|xmpp):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i;
var clean = DOMPurify.sanitize(dirty, {ALLOWED_URI_REGEXP: /^(?:(?:(?:f|ht)tps?|mailto|tel|callto|cid|xmpp|xxx):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i;});
```

so to allow file protocol, the REGEX would need to be something like:
_/^(?:(?:(?:f|ht)tps?|mailto|tel|callto|cid|xmpp|**file**):|[^a-z]|[a-z+.\-]+(?:[^a-z+.\-:]|$))/i;_

**Comment by davidjgraph** [comment]
Thanks, we're aware of the problem and we're looking into it.

**Comment by adamkingsbury** [comment]
hi @davidjgraph I should also point out that this bug affect ALL versions of draw.io retrospectively as the exported HTML uses a script link to "https://viewer.diagrams.net/js/viewer-static.min.js"

as such all previous versions of exported HTML from draw.io are affected by the changeover to DOMPurify, effectively retrospectively breaking all previously working exports.

I just tested versions 18 through to 13 before realising.

Hoping that may raise the priority a bit.

**Comment by davidjgraph** [comment]
Yes. The reason for the change is we had a number of security vulnerabilities reported and the regex had to be heavily tighten as security is the priority. We're treating this ticket as the issue with relative links, which I don't see a security issue with, currently. We need to consider the desktop tool, too, though.

File protocol links would need another issue, they are not so simple.

**Comment by cgulati001** [comment]
@adamkingsbury, I found that the embedded links in the exported draw.io HTMLs work on a hosted server just not in the local environment. The problem I am now facing is how to test new views when development is done on a local machine.

**Comment by adamkingsbury** [comment]
@cgulati001 

> If anyone can share a workaround I would appreciate it. Thank you.

I have found and tested a workaround solution assuming you are exporting the html and accessing the file from a network drive or local PC (not deployed over web server....though this could potentially work for that too)

1. I downloaded the release source code for [draw.io v17.5.1](https://github.com/jgraph/drawio/releases/tag/v17.5.1). This was the last release before DOMpurify was included.
2. I extracted 2 files from **drawio-17.5.1.zip\drawio-17.5.1\src\main\webapp\js**
  - viewer.min.js
  - viewer-static.min.js

3. I added a new folder in the same location as the Exported HTML file is located. In my example I called the new folder "tempfix" and added the extracted files to the folder. Now my file structure looks like this:

```
Root Folder
|--processmap.drawio
|--processmap.html
|--processes
|   |--referencedfile.html
|   |--referencedfile2.html
|
|--tempfix
    |-- viewer.min.js
    |-- viewer-static.min.js
```

4) I then edited the exported html changing this line:
`<script type="text/javascript" src="https://viewer.diagrams.net/js/viewer-static.min.js"></script>`

to this:
`<script type="text/javascript" src="tempfix/viewer-static.min.js"></script>`

You will need to re-edit the script link if you re-export the html file again from draw.io, but you can also fix any previously working exports using this workaround too.

**Comment by adamkingsbury** [comment]
> @adamkingsbury, Thank you Adam I will try this. Do I need to uninstall the current draw.io version first?

No, you can keep any version of the desktop client

**Comment by cgulati001** [comment]
@adamkingsbury, I was able to follow your workaround until Step 4. Where are you making this line change?

**Comment by adamkingsbury** [comment]
> @adamkingsbury, I was able to follow your workaround until Step 4. Where are you making this line change?

in step 4 I edited **processmap.html**....the html file that was exported from draw.io

**Comment by adamkingsbury** [comment]
The only other suggestion I have for you is that notepad++ (open source
text editor) has the ability to perform a text find and replace over all
files in a directory. You could use this as a way to make the correction in
bulk. here's a tutorial
<https://darrenoneill.eu/?p=1250#:~:text=Notepad%2B%2B%20allows%20you%20to,in%20multiple%20files%20are%20once.&text=First%20open%20Notepad%2B%2B%20and,then%20click%20find%20in%20files.&text=Click%20ok%20to%20allow%20the,how%20many%20occurrences%20were%20replaced.>

Otherwise it'll be a waiting game for a resolution to the issue from the
draw.io team.

**Comment by davidjgraph** [comment]
Fixed in 18.1.3

**Comment by cgulati001** [comment]
@davidjgraph Thank you!!!!!
